### PR TITLE
Fixed logic bug in JPath::isOwner

### PR DESCRIPTION
**Pull request by SniperSister**
## Issue

There's a logic bug in JPath::isOwner causing false results of the ownership check. Let's assume:
- /tmp is unwriteable
- session.save_path is writeable

Expected result: session.save_path is used as $dir

Actual result: As $dir is a string, !$dir becomes a boolean false and the whole check returns false. This causes that the correct $dir is overwritten with a false in line 241.

I found this issue because I was trying to save a configuration.php with chmod 444 on a webspace with an unwriteable /tmp and a writeable session.save_path. Because of the logic bug, isOwner doesn't return true as expected and Joomla fails to execute a chmod 644 which would allow us to update the file.
## Solution

I fixed the logic bug and improved the readablity by converting the code into a foreach loop.
## Testing

That's a bit harder. You need a server with:
- /tmp unwriteable for PHP (i.e. because it's not included in the open_basedir)
- session.save_path writeable

If you have such a machine. Change the chmod of the configuration.php to 444 and try to save the global configuration. You'll get an errror message.

Apply the patch, try to save the configuration.php again, now it works as expected.

### CODE REVIEW

File: libraries/joomla/filesystem/path.php [hunk 1]
```diff
@@ -236,9 +236,17 @@ public static function isOwner($path)
 		$jtp = JPATH_SITE . '/tmp';
 
 		// Try to find a writable directory
-		$dir = is_writable('/tmp') ? '/tmp' : false;
-		$dir = (!$dir && is_writable($ssp)) ? $ssp : false;
-		$dir = (!$dir && is_writable($jtp)) ? $jtp : false;
+		$dir = false;
+
+		foreach (array($jtp, $ssp, '/tmp') as $currentDir)
```
Comment (wilsonge) [inline libraries/joomla/filesystem/path.php]: We now appear to be checking in the reverse order to present. Surely we should be checking `/tmp` first then `$ssp` then `$jtp`.

**Comment by wilsonge** [comment]
OK so obvious question maybe but why is `$dir` a string in the first place? The only way that can happen is if `is_writable('/tmp')` returns `true` which suggests you have bigger problems going on?

File: libraries/joomla/filesystem/path.php (hunk 1 above)
Comment (SniperSister) [inline libraries/joomla/filesystem/path.php]: @bakual suggested this as $jtp is the most likely writeable directory

**Comment by SniperSister** [comment]
I don't get your last comment @wilsonge, can you rephrase it?

File: libraries/joomla/filesystem/path.php (hunk 1 above)
Comment (Bakual) [inline libraries/joomla/filesystem/path.php]: We discussed that on Glip and we thought it makes most sense if the temp directory in the Joomla folder is tried (and used) first. With falling back to the session temp folder and then to a (maybe) existing tmp directory in server root.

**Comment by Bakual** [comment]
@SniperSister I think the current code is flawed that much that George has the same issues understanding it like I had yesterday in Glip :smile:

**Comment by wilsonge** [comment]
No I see the issue now. What that code should have been was

```
$dir = is_writable('/tmp') ? '/tmp' : false;
$dir = (!$dir && is_writable($ssp)) ? $ssp : $dir;
$dir = (!$dir && is_writable($jtp)) ? $jtp : dir;
```

Now it makes sense why you are changing the order as well - because currently the other two are completely ignored anyhow.

**Comment by ChristineWk** [comment]
Hi David,
confirm: Actual result: As $dir is a string, !$dir becomes a boolean false and the whole check returns false. This causes that the correct $dir is overwritten with a false in line 241. https://github.com/joomla/joomla-cms/commit/74db56fdfe612609caa71c650f25140811de7c74

Confirm integrated deletions/additions to /libraries/joomla/filesystem/path.php and the patch solves it.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8532">issues.joomla.org/joomla-cms/8532</a>.</sub>

**Comment by ChristineWk** [comment]
I have tested this item :white_check_mark: successfully on 86faeefd1b0a8b9bec4e1fd999064dd4ad1826e0<br /><br />confirm patch successfully. <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8532">issues.joomla.org/joomla-cms/8532</a>.</sub>

**Comment by anibalsanchez** [comment]
I have tested this item :white_check_mark: successfully on 86faeefd1b0a8b9bec4e1fd999064dd4ad1826e0<br /><br />Test OK, configuration.php can be saved and it remains 444<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8532">issues.joomla.org/joomla-cms/8532</a>.</sub>

**Comment by SniperSister** [comment]
Awesome, thanks folks! Setting to RTC as we're having 2 successful tests.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8532">issues.joomla.org/joomla-cms/8532</a>.</sub>

**Comment by richard67** [comment]
@wilsonge @SniperSister Any idea how it will go on with this PR? It is RTC but has no milestone. If it shall be merged, I suggest to change it to correct the issue in the same way as @wilsonge did for the framework with commit [5ee2159](https://github.com/joomla-framework/filesystem/commit/5ee2159d3e304fb367696951e68dc05466de90d6 "joomla-framework - commit 5ee2159"). Or is it solved with the commit on the framework for the CMS, too, means the file libraries/joomla/filesystem/path.php is taken from the framework? In this case this PR here should be closed, or not?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8532">issues.joomla.org/joomla-cms/8532</a>.</sub>

**Comment by wilsonge** [comment]
Checking tmp first does make sense. This will do for now.
